Test Books Controller

cquiro opened the pull request:
## Summary

* Create spec/support/shared_contexts/authenticated_user.rb
* Include Devise::Test::ControllerHelpers in rails_helper.rb
* Write specs for BooksController#show and BooksController#index

## Trello Card

https://trello.com/c/LYQOznHR/24-crear-tests-para-rutas-index-y-show-de-book

cquiro committed:
Create authenticated_user.rb shared context.

cquiro committed:
Merge branch 'master' into test_books_controller

cquiro committed:
Write BooksController specs

- Write passing specs for BooksController#show

- Write failing specs for BooksController#index

- Add devise controller helpers in rails_helper.rb

cquiro committed:
Merge branch 'show_index_books' into test-books-controller

cquiro committed:
Move BookSerializer out of Api::V1

- Move BookSerializer out of api/v1

- Replace Api::V1::BookSerializer with BookSerializer in every file.

- Add Api::V1 namespace to ApiController

cquiro committed:
Merge branch show_index_books into test-books-controller

lmasello commented on spec/controllers/api/v1/books_controller_spec.rb:
```diff
@@ -0,0 +1,38 @@
+  describe 'GET #index' do
+    let!(:books) { create_list(:book, 3) }
+
+    before do
```
You can do it in one line too :)

facucompe commented on app/controllers/api/v1/books_controller.rb:
```diff
@@ -2,11 +2,11 @@ module Api
   module V1
     class BooksController < ApiController
       def index
-        render_paginated books, each_serializer: Api::V1::BookSerializer, status: :ok
```
Hahah, you correct it here. Nice!

facucompe commented on spec/controllers/api/v1/books_controller_spec.rb:
```diff
@@ -0,0 +1,38 @@
+      expected = ActiveModel::Serializer::CollectionSerializer.new(
+        books, each_serializer: BookSerializer, root: false
+      ).to_json
+      expect(response.body.to_json) =~ JSON.parse(expected)
```
In the example the test says `response_body`. We usually have this helper for tests to parse the response body.
```
module Response
  module JSONParser
    def response_body
      ActiveSupport::JSON.decode(response.body) if response.present? && response.body.present?
    end
  end
end
```
Add it in **spec/support**

facucompe replied:
Also add  `config.include Response::JSONParser, type: :controller` to your rails_helper

cquiro replied:
@facucompe Done!

facucompe reviewed (changes requested):
God Job!

cquiro committed:
Merge branch 'show_index_books' into test-books-controller

cquiro committed:
Write before blocks as one-liners in books_controller_spec.rb

cquiro committed:
Add response_body helper method

- Create response_helper.rb file with a
Response::JSONParser.response_body method

- Include Response::JSONParser in rails_helper.rb (Rspec.configure)

- Use response_body method in books_controller_spec.rb
